Plans: add main plan component — PlanFeatures

lamosty opened the pull request:
Add `PlanFeatures` component as a part of bigger plans-redesign started in #5715.

## Example usage

```js
import PlanFeatures from 'my-sites/plan-features';
import { PLAN_FREE } from 'lib/plans/constants';
import QueryPlans from 'components/data/query-plans';

export default MyComponent() =>
   <div>
      <QueryPlans />
      <PlanFeatures plan={ PLAN_FREE } />
   </div>
```

## Testing

1. `PlanFeatures` was added to devdocs into App Components;
2. Navigate to http://calypso.localhost:3000/devdocs/app-components/plan-features;
3. See the redesigned plan components (free, premium and business);
4. Look for errors;
5. Is everything fine?

## Screenshots

![selection_022](https://cloud.githubusercontent.com/assets/4988512/16093020/726d3ef0-333a-11e6-9d47-8d5160a9cb96.jpg)

![selection_023](https://cloud.githubusercontent.com/assets/4988512/16093030/7eeb6ec2-333a-11e6-9274-4796f90bab43.jpg)

**Note:** In the second screenshot, the Business plan has a strange price. This is probably related to the recent PR #6001 by @gwwar but she's already working on a better solution here: #6046. That's why I've left it as it is.

If it is not related to that, I'll fix it.

Although this is still work in progress, I think I've arrived to a presentable code so please review. :)

cc @gwwar @rralian @mtias @artpi @retrofox 

Test live: https://calypso.live/?branch=add/plan-features-plan

gwwar commented on client/lib/plans/constants.js:
```diff
@@ -118,3 +182,11 @@ export const getPlanObject = planName => {
+		obj.plans.indexOf( planName ) !== -1
+	);
+
+	return sortBy( planFeatures, 'order' );
```
I understand what you're doing here, but this may be somewhat difficult to maintain in the future. Maybe add a list of feature references to the plan object instead?

e.g.
```javascript
	[ PLAN_FREE ]: {
        getFeatures: () => [ FEATURE_3GB_STORAGE, FEATURE_WP_SUBDOMAIN,  ... ]
		getTitle: () => i18n.translate( 'Free' ),
		getPriceTitle: () => i18n.translate( 'Free for life' ),
		getProductId: () => 1,
		getDescription: () => i18n.translate( 'Get a free blog and be on your way to publishing your first post in less than five minutes.' )
	},
```

lamosty replied:
Hmm, I thought we don't want to duplicate the references (as they are currently made from the opposite direction). But I'll implement this as I think it's a better solution. We also don't need the `order` prop in features because we can just order them in the array.

gwwar commented on client/my-sites/plan-features/style.scss:
```diff
@@ -1,3 +1,29 @@
+	background-color: lighten( $gray, 35% );
+
+	&.is-popular {
+		margin-top: 35px; // popular banner + margin
```
Maybe add some local sass variables, so this will be updated automatically when values height or margin change.

gwwar commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	return {
+		planName: ownProps.plan,
```
I would assume ownProps.plan is a plan object. Maybe passing through planName would be better.

gwwar replied:
Actually on second thought, I think I'm fine with the name

gwwar commented on client/my-sites/plan-features/index.jsx:
```diff
@@ -0,0 +1,78 @@
+
+export default connect( ( state, ownProps ) => {
+	const planProductId = plansList[ ownProps.plan ].getProductId();
+	let sitePlanObj = getCurrentPlan( state, sitesFactory().getSelectedSite() );
```
There's also a `getSelectedSiteId` selector available in `state/ui/selectors` which lets us avoid using SitesList directly

gwwar replied:
If the selected site is populated, we may be able to avoid needing to populate the sites/plans endpoint that getCurrentPlan depends on.

The site object has a plan property which looks like:
![screen shot 2016-06-15 at 4 26 36 pm](https://cloud.githubusercontent.com/assets/1270189/16100829/f6715d96-3315-11e6-9ba0-c5e5dfe2b10c.png)

We could add a new selector like `getSitePlan` in `sites/selectors`

lamosty replied:
Great idea. I'll add the `getSitePlan` selector in another PR and rebase this one on top of it.

gwwar commented:
👍  Looking good so far. #6046 should fix the visual oddities with price. Let me know if you need help with new selectors.

lamosty commented:
@gwwar Thanks for all the comments.

I've added the new selectors (in #6082) and rebased this PR on top of that PR so please ignore the noise. Does it look better? :)

gwwar commented on client/lib/plans/constants.js:
```diff
@@ -105,6 +148,41 @@ export const featuresList = {
+
+	[ FEATURE_WP_SUBDOMAIN ]: {
+		getTitle: () => i18n.translate( 'WordPress.com subdomain' ),
+		plans: [ PLAN_FREE ]
```
Hmm this could get out of sync too. If we need the reverse lookup maybe lets remove this, and add in helper function like:

```javascript
getPlansForFeature( feature )
```

We could also memoize the results if this is called frequently

lamosty replied:
Good point. Do we want to do it in this PR?

gwwar replied:
We can leave this for a future PR

gwwar commented on client/my-sites/plan-features/style.scss:
```diff
@@ -1,3 +1,33 @@
+$header-banner-height: 20px;
```
lets move this under the .plan-features scope so the sass variable name is not global

lamosty replied:
Oops, I wasn't sure whether it gets scoped locally or globally if I did that. Good to know, thanks.

lamosty replied:
> lets move this under the .plan-features scope so the sass variable name is not global

I'm just trying to make it work but when a Sass var is defined in a scope, it can't be used in another scope. In Calypso, we don't write CSS selectors like `.plan-features .header-banner` but like `.plan-features__header-banner`, therefore putting `$header-banner-height` under `.plan-features` scope won't work (unless I'm missing some Calypso magic :sparkles: ).

![selection_019](https://cloud.githubusercontent.com/assets/4988512/16147877/4804ec10-3486-11e6-86e4-481258999dc4.jpg)

gwwar replied:
Good Point. Lets namespace the variable like `$plan-features-header-banner-height` or you can just use the bare value. Feel free to choose what you think works better.

gwwar commented:
Looking good! I left a few notes, but when you're ready, this is good to 🚢

lamosty commented:
@gwwar I've rebased this PR on top of master but I'm getting errors (screenshots below) now I wasn't getting before. I started rebasing this on Friday (so the first rebase includes your #6046). After this rebase, I think I got no new errors. However, I then pulled in new code in `master` today, rebased it again and now I'm getting the errors (screenshots below).

![selection_026](https://cloud.githubusercontent.com/assets/4988512/16198612/cb7cb032-3706-11e6-9aad-6ea878d7d32c.jpg)

The first part about state validation failed is a bit unknown to me (I only know that we store state to localForage and on loading it back we validate it — but I don't know what I should do with this error). The second part refers to `my-sites/plan-features/index.jsx`. I'm getting `bill_period_label` out of a plan object, like this: 

```js
billingTimeFrame: getPlan( state, planProductId ).bill_period_label
```
However, initially when there are no plans data, it probably fails because `undefined.bill_period_label` is not a thing (of course). Shouldn't we add the loading placeholder plan like it's currently being done on the `/plans` page?

gwwar commented:
:shipit:
